lib/read.js

```diff
@@ -37,24 +39,45 @@ module.exports = class PromiseReadStream extends ReadableStream {
       Promise.resolve()
         .then(() => readCb.call(this, bytes))
         .then(data => {
-          this._reading = false
           if (data !== undefined) {
             this.push(data)
           }
+          return Promise.all(this._pushQueue)
+        })
+        .then(() => {
+          this._reading = false
           if (this._keepReading) {
             this._read()
           }
-        }, this.emitError)
+        })
+        .catch(err => this.emitError(err))
     }
   }
 
   push (data) {
-    return Promise.resolve(data)
+    if (data === null) {
+      return this._endingPush()
+    }
+    const readOperation = Promise.resolve(data)
       .then(data => {
-        this._keepReading = super.push(data)
-      }, err => {
-        this.emitError(err)
+        this._pushQueue.delete(readOperation)
+        if (data === null) {
+          return this._endingPush()
+        }
+        this._keepReading = super.push(data) && !this._ending
       })
+      .catch(err => this.emitError(err))
+    this._pushQueue.add(readOperation)
+    return readOperation
+  }
+
+  _endingPush () {
+    this._keepReading = false
+    this._ending = true
+    return Promise
+      .all(this._pushQueue)
+      .then(() => super.push(null))
```

Comment:
This is the one line that is tripping me up. It looks like we push `null` whenever a push is done or is this line `.all(this._pushQueue)` dynamic and the size of `this._pushQueue` growing as the read stream reads from it's source?